lang/parser.y

```diff
@@ -1179,16 +1179,18 @@ type:
 		names := make(map[string]struct{})
 		strs := []string{}
 		for _, arg := range $3.args {
-			s := fmt.Sprintf("%s %s", arg.Name, arg.Type.String())
-			if _, exists := names[s]; exists {
+			field := fmt.Sprintf("%s %s", arg.Name, arg.Type.String())
+			if _, exists := names[arg.Name]; exists {
 				// duplicate field name used
-				err := fmt.Errorf("duplicate struct field of `%s`", s)
+				err := fmt.Errorf("duplicate struct field of `%s`", field)
 				// this will ultimately cause a parser error to occur...
 				yylex.Error(fmt.Sprintf("%s: %+v", ErrParseSetType, err))
-				break // we must skip, because code continues!
+				// we must immediately get out of here to avoid other parts
+				// of the parser from using this type here.
+				goto ret1
```

Comment:
I'm not convinced this is appropriate-- it means we're depending on something in goyacc that could be unstable. Even without this, it should work fine.

Reply:
Hu, seems like I just forgot to `make lang` before retesting this, it actually _does_ work with the break in place, so I have removed my change for now.

That said I do get a few hundred execs per second less in go-fuzz using break vs return 1.  I find both goyacc as well as the plan9 yacc man page lacking, but I do think returning non zero ints anywhere in the parser or lexer should be fine to signal failed parsing/lexing. Is that correct? Why not just do that?

Reply:
> That said, why not just return 1 after any call to yylex.Error? I find both goyacc as well as the plan9 yacc man page lacking, but I do think returning non zero ints anywhere in the parser or lexer should be fine to signal failed parsing/lexing. Is that correct?

It's a good question. I don't know the answer, but I found the yylex.Error somewhere I believe in some sample code somewhere, so I got the impression it was correct like that. If I would *guess*, here's what I would say:

Letting the parser "finish" gives it a chance to potentially find a more serious problem, and/or a second problem, and in theory, display multiple useful errors. I'm not sure it currently can do this, but it might be somewhere we could get to.